### ISSUE TITLE
Appropriate Benchmarking

### ISSUE THREAD

**Issue by asistradition** [morris-lab/CellOracle#58]
Good afternoon-

I'm benchmarking the performance of this network inference tool and I'd just like to double-check that the workflow that I have is appropriate to generate a GRN from single-cell data and an externally-provided base GRN  (I'm using the tutorial as a guideline):

1. Filter genes with no counts
2. Normalize per-cell depth
3. Filter to 2000 highest-variance genes (is this strictly necessary? I'd prefer to model all of them)
4. Renormalize per-cell depth based on filtered genes
5. Log1p & scale data
6. PCA and kNN graph (k=4 on 20 PCs)
7. Calculate diffusion map
8. kNN graph (k=10 on diffusion components)
9. Louvain clustering on diffusion kNN (resolution=0.8)
10. Dimensional reduction with PAGA and FORCE 
11. KNN Imputation of raw data with ORACLE (k and n_comps determined with heuristic)
12. Infer GRN links 

Code follows:
```
sc.pp.filter_genes(adata, min_counts=1)
sc.pp.normalize_per_cell(adata, key_n_counts='n_counts_all')
filter_result = sc.pp.filter_genes_dispersion(adata.X, flavor='cell_ranger', n_top_genes=2000, log=False)
adata = adata[:, filter_result.gene_subset]
sc.pp.normalize_per_cell(adata)

adata.raw = adata
adata.layers["raw_count"] = adata.raw.X.copy()

sc.pp.log1p(adata)
sc.pp.scale(adata)

sc.tl.pca(adata, svd_solver='arpack')
sc.pp.neighbors(adata, n_neighbors=4, n_pcs=20)
sc.tl.diffmap(adata)
sc.pp.neighbors(adata, n_neighbors=10, use_rep='X_diffmap')
sc.tl.louvain(adata, resolution=0.8)

sc.pl.paga(adata)
sc.tl.draw_graph(adata, init_pos='paga', random_state=123)

sc.pl.draw_graph(adata, color='louvain', legend_loc='on data')

oracle.import_anndata_as_raw_count(adata=adata,
                                   cluster_column_name="louvain",
                                   embedding_name="X_draw_graph_fa")

oracle.addTFinfo_dictionary(base_GRN_TG_to_TF)

n_comps = np.where(np.diff(np.diff(np.cumsum(oracle.pca.explained_variance_ratio_))>0.002))[0][0]
n_cell = oracle.adata.shape[0]
k = int(0.025*n_cell)

oracle.knn_imputation(n_pca_dims=n_comps, k=k, balanced=True, b_sight=k*8, b_maxl=k*4, n_jobs=4)

links = oracle.get_links(cluster_name_for_GRN_unit="louvain", alpha=10,
                         verbose_level=10, test_mode=False)
```
Thanks for your time

**Comment by KenjiKamimoto-wustl122** [comment]
@asistradition 

Thank you for using CellOracle.
I'm Kenji, a developer of CellOracle.

Yes, I think you are following the workflows correctly. 
FYI, Let me make some suggestions as follows.

**Suggestion 1)**
The workflow in this tutorial is primarily made for the perturbation simulation. 
If you are solely interested in GRN inference and not interested in the simulation, you can change some steps.

1-1) You can change the gene filtering parameter if you do not do perturbation simulation.
Although we recommend doing gene filtering in our tutorial, the aim of this gene filtering is to improve simulation results.
I guess you would like to use all genes for GRN inference benchmarking. You can skip the gene filtering step below in this context.
```
filter_result = sc.pp.filter_genes_dispersion(adata.X, flavor='cell_ranger', n_top_genes=2000, log=False)
adata = adata[:, filter_result.gene_subset]
```

1-2) 
In the workflow in this tutorial, we calculate dimensional reduction, but it is only required in the simulation step.
The diffusion map, PAGA analysis, and force-directed network analysis results do not affect GRN inference results.
You can skip this process.
For example, you can skip these processes below,
```
sc.pl.paga(adata)
sc.tl.draw_graph(adata, init_pos='paga', random_state=123)
sc.pl.draw_graph(adata, color='louvain', legend_loc='on data')
```
Instead of them, please use PCA in the later stage.
```
oracle.import_anndata_as_raw_count(adata=adata,
                                   cluster_column_name="louvain",
                                   embedding_name="X_pca")
```

1-3)
In this workflow, we need to do a knn imputation for the simulation step. It is also not necessary for GRN inference.
Suppose you want to use gene expression data without any imputation process. In that case, you can substitute gene expression value with non-imputed value by running the following command between knn imputation and GRN calculation. 
```
oracle.adata.layers["imputed_count"] = oracle.adata.layers["normalized_count"].copy()
```



**Suggestion2)**

As described above, the CellOracle workflow we introduced in our tutorial were designed to do perturbation simulation.
If you are interested in GRN calculation solely, the notebook below might be a good option.
I write a minimum script for GRN analysis. In this notebook, we use the `Net` class, a core algorithm used in the `Oracle` object. 
https://github.com/morris-lab/CellOracle/blob/master/docs/notebooks/Net_class/scanpy_preprocessing_with_Paul_etal_2015_data.ipynb

I hope it helps.

Kenji

**Comment by asistradition** [comment]
That's very helpful, thank you very much. One small followup question:

I'm ranking the output network edges by -logp, but there's a small number of edges reported with p values that are NaN and coef_mean of 0. I was planning to drop them, but I wanted to check if there's a reason to keep them.

Thanks again for your help

**Comment by KenjiKamimoto-wustl122** [comment]
@asistradition 

As you said, I think you can drop them. 
This phenomenon may happen in some cases. 

**Case 1) Gene expression value is zero in a cluster.**
CellOracle constructs a GRN model for each cluster. Even if we removed zero expression genes in the scRNA-seq preprocessing step, all gene expression values can be zero in a cluster.
If so, CellOracle cannot construct a GRN model for this gene, generating coef_meam=0 and p_val = Nan.

**Case 2) Bagging** 
The bagging function may be a cause of this issue. But very unlikely in most cases.
Please read the following explanation if you are interested in the detail of the CellOracle algorithm. 
I provide this information FYI, but I guess It is not related to your issue.

CellOracle makes bagging models In the GRN calculation. It uses sckit-learn `BaggingRegressor` function.
Here is the source code. Please look scikit-learn web page together. 
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingRegressor.html
```
model = BaggingRegressor(base_estimator=Ridge(alpha=alpha,
                                                solver=solver),
                                                n_estimators=bagging_number,
                                                bootstrap=True,
                                                max_features=0.8,
                                                n_jobs=n_jobs,
                                                verbose=False,
                                                random_state=123)
```
As shown above, we set max_features=0.8. It means we randomly pick up and use 80% of regulatory genes in one ML model, but we repeat this calculation many times. 
If you set a very small bagging number in the GRN calculation, we can fail to analyze some genes. For example, if we set bagging_number=1 in the `get_links` function, we cannot analyze 20% of the regulatory gene.
https://morris-lab.github.io/CellOracle.documentation/modules/celloracle.html#celloracle.Oracle.get_links
The default bagging number is 20. This is not that big, but we usually do not get p=Nan with this bagging number 20.


I think case1 is most likely, and I guess you can drop such a gene from inferred GRN in this case.
But if you find CellOracle behaves unexpected/strange behavior, please let me know. I will check our codes.

Kenji

**Comment by asistradition** [comment]
Given the data that I'm feeding in, gene expression of zero in a cluster is very reasonable (I'm using a data set that's got several very different clusters in it), so that makes sense.

I've gotten everything working nicely I think (the AUPR when scored against my gold standard without holdouts is ~0.5, which is reasonable). I just want to make sure that I'm interpreting my results correctly; the bagging regression will only consider network edges which were in base_GRN, correct? If there are no edges for a gene in base_GRN, then the GRN inference will always return no edges in the learned GRN?

**Comment by KenjiKamimoto-wustl122** [comment]
@asistradition 

Sorry for the delayed response.

>> the bagging regression will only consider network edges which were in base_GRN, correct? If there are no edges for a gene in base_GRN, then the GRN inference will always return no edges in the learned GRN?

Yes, exactly.  It only consider the network edge in the base-GRN.
